044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,270 @@
+connect to the replication listener and will piggyback on the broker's existing SSL configuration to
+obtain keystore/truststore properties.
+
+If using cluster sourced volumes we will expose these optional properties to customise the admin connection:
+
+1. `client.quota.callback.kafka.listener.name` (default `replication-9091`)
+2. `client.quota.callback.kafka.listener.port` (default `9091`)
+3. `client.quota.callback.kafka.listener.protocol` (default `SSL`)
+4. `client.quota.callback.kafka.clientIdPrefix` (default `__strimzi`)
+
+Setting `client.quota.callback.static.storage.hard` or `client.quota.callback.static.storage.soft` would be incompatible
+with the cluster sourced volumes and fail configuration of the plugin.
+
+The admin client will re-use the broker SSL configuration
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+used bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume.
+
+We retain the existing concept of a soft limit (where throttling begins) and a hard limit (where we effectively
+stop message production) for all limit types.
+
+The only requirement is a single hard-limit configuration, if no soft-limit is configured we will infer that soft-limit==hard-limit.
+
+The different limit types can be combined, so you could combine a usedBytesExceeds soft limit with a availableRatioBelow
+hard limit.
+
+Defining multiple soft limits or multiple hard limits would be an invalid state.
+
+The limits we want are:
+1. throttle if [usedBytes](#volume) is greater-than-or-equal-to some threshold for any volume
+2. throttle if [availableBytes](#volume) is less-than-or-equal-to some threshold for any volume
+3. throttle if the [availableRatio](#volume) is less-than-or-equal-to some threshold for any volume
+
+Using configuration properties like:
+
+1. `client.quota.callback.static.storage.perVolumeLimit.usedBytesExceeds.soft=14000000000`
+2. `client.quota.callback.static.storage.perVolumeLimit.usedBytesExceeds.hard=15000000000`
+3. `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.soft=5000000000`
+4. `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=1000000000`
+5. `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+6. `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.hard=0.01`
+
+`availableRatioBelow` value can be in the range (0.0, 1.0)
+
+#### Fallback Throttle Factor
```

Comment:
In discussion today it came up that in our fork we have a concept of 'snapshot' which is intended to make this a little more resilient.

Maybe we should add something similar. The idea being that we can operate using an older, valid set of observations, up to some time limit.

for example:
1. at 13:05 the plugin makes a valid observation of the cluster and calculate throttle factor 1.0, we use that and record the time
2. at 13:06 the plugin tries and fails to observe the cluster's state, but we decide that the previous observations are still fresh enough so we continue with 1.0
3. at 13:07 the plugin tried and fails to observe the cluster's state, we decide the previous observations are no longer fresh enough, so we use the fallback throttled factor
4. at 13:08 the plugin makes a valid observation of the cluster and calculate throttle factor 1.0, we use that and record the time

Reply:
I can see why falling back to last known value would make sense. Not entirely sure why to expire it - why would it not make sense until you have new data instead of the fallback factor.

Reply:
The fallback factor would be applied in two scenarios:
1. Cold start, the plug-in needs a factor to apply until it can get usage data from the admin API.
2. Transient consistency issues. The plug-in identifies the active brokers to be `[1,2,3]` but can only get usage data for `[1,3]`. 

A fallback factor addresses both problems. 

Caching the usage data/ previous factor solves only 2. The trouble with caching is the accuracy of the data decays with time, and the rate of decay varies between users. That said I think the caching is cheap to do and provides a better experience for users so is worth having. 

I'm thinking it is better to define the the expiry in multiples of the check interval rather than as an explicit time, as the cache is intended to cover transient problems rather than long term outages. By offering control of the cache expiry users could set effectively infinite expiry (i.e. `1,000,000` check intervals).

Reply:
And now reworked to a `Throttle Factor Validity Period` to disconnect it from the recalculation implementation, make it less surprising and avoid the kafka 'poll' language.